building/devices/versal_net.rst

```diff
@@ -0,0 +1,555 @@
+	* TEE_ERROR_GENERIC - PLM failure
+	*/
+	#define PTA_VERSAL_FPGA_WRITE		0x0
+
+	TEE_Result load_bitsream(uint8_t *bistream, size_t size)
+	{
+		TEEC_Context ctx;
+		TEEC_Session sess;
+		TEEC_Operation op;
+		TEEC_UUID uuid = PTA_VERSAL_FPGA_UUID;
+		TEE_Result ret = TEE_SUCCESS;
+		uint32_t origin;
+
+		ret = TEEC_InitializeContext(NULL, &ctx);
+		if (ret != TEEC_SUCCESS)
+			return ret;
+
+		/* Open a session with the TA */
+		ret = TEEC_OpenSession(&ctx, &sess, &uuid,
+			       TEEC_LOGIN_PUBLIC, NULL, NULL, &origin);
+		if (ret != TEEC_SUCCESS)
+			goto out;
+
+		memset(&op, 0, sizeof(op));
+		op.paramTypes = TEEC_PARAM_TYPES(TEEC_MEMREF_TEMP_INPUT,
+						 TEEC_NONE, TEEC_NONE, TEEC_NONE);
+
+		op.params[0].tmpref.buffer = bitstream;
+		op.params[0].tmpref.size = size;
+
+		ret = TEEC_InvokeCommand(&sess, PTA_VERSAL_FPGA_WRITE,
+					 &op, &origin);
+
+		TEEC_CloseSession(&sess);
+	out:
+		TEEC_FinalizeContext(&ctx);
+		return ret;
+	}
+
+The example above shows a ``load_bitsream()`` function that expects a buffer
+(and its size) as a parameter. This buffer holds the actual bitsream binary
+loaded from Linux filesystem for instance.
+
+.. note::
+	Bitsreams loaded through this means have their size limited by
+	the amount of shared memory available to OP-TEE. Bigger bitsreams
+	should be loaded at boot time.
+
+NVM
+===
+
+The Versal Net OP-TEE provides eFuses read and write APIs to other OP-TEE
+components. The API is available in ``core/include/drivers/versal_nvm.h``.
+
+.. code-block:: c
+	:caption: Example - Read the DNA value
+
+	#include <drivers/versal_nvm.h>
+
+	TEE_Result read_dna(uint32_t *dna)
+	{
+		return versal_efuse_read_dna(dna, EFUSE_DNA_LEN);
+	}
+
+.. code-block:: c
+	:caption: Example - Write Black Obfuscation IV
+
+	#include <drivers/versal_nvm.h>
+
+	TEE_Result write_black_iv(uint32_t *iv)
+	{
+		struct versal_efuse_ivs ivs = { };
+
+		ivs.prgm_blk_obfus_iv = 1;
+		memcpy(ivs.blk_obfus_iv, iv, EFUSE_IV_LEN);
+
+		return versal_efuse_write_iv(&ivs);
+	}
+
+PUF
+===
+
+The Versal Net Physically Unclonable Function is support on the OP-TEE port.
+The API is available in ``core/include/drivers/versal_puf.h``.
+
+.. code-block:: c
+	:caption: Example - PUF Registration
+
+	#include <drivers/versal_puf.h>
+
+	TEE_Result register_puf(struct versal_puf_data *data)
+	{
+		struct versal_puf_cfg cfg = { };
+
+		cfg.puf_operation = VERSAL_PUF_REGISTRATION;
+		cfg.shutter_value = VERSAL_PUF_SHUTTER_VALUE;
+		cfg.global_var_filter = VERSAL_PUF_GLBL_VAR_FLTR_OPTION;
+		cfg.read_option = VERSAL_PUF_READ_FROM_RAM;
+
+		return versal_puf_register(data, &cfg);
+	}
+
+Testing
+*******
+
+GPIO
+====
+
+[Tracked by requirement R-4]
+
+.. note::
+	Fully testing this requires plugging an actual load on the corresponding GPIO pin.
+
+	The pin used can be modified in ``core/pta/versal/test_pta.c`` by changing ``GPIO_TEST_PIN_ID``.
+
+This test is available in the ``versal`` testsuite in ``xtest``:
```

Comment:
I can't find that in xtest.

Reply:
These tests won't be upstreamed. This section of the doc was mostly AMD/Xilinx internal discussion. I removed it.